### PR TITLE
Experimental formatter for bzl files with def expressions

### PR DESCRIPTION
**Pull request by vladmos**

### CODE REVIEW

**Comment by vladmos** [comment]
This pull request doesn't provide a complete support of bzl files, only def-blocks are currently supported, however it already supports nested structures, so adding for-, if- and if-else-blocks is just a matter of adding a couple of new nodes.

Currently known problems (to be fixed in the follow-up pull requests):

  * Tabs are not supported properly,
  * If a function definition contains a comment it will be moved to its last argument (that's reflected in the 051.formatbzl.golden testfile)

The behavior shouldn't be changed if the `--format_bzl` flag is not passed explicitly, so these limitations shouldn't actually break anything.

File: build/lex.go [hunk 1]
```diff
@@ -571,13 +632,19 @@ func (in *input) skipPython(p []byte) []byte {
...
 			}
 
-			if !isBlankOrComment(p[i:]) {
-				if !hasPythonContinuation(p[i:]) && c != ' ' && c != '\t' {
-					// Yes, stop here.
-					break
+			if tables.FormatBzlFiles {
+				// In the bzl files mode we only care about the end of the statement, we've found it.
+				break
+			} else {
```
Comment (pmbethe09) [inline build/lex.go]: don't need 'else' when the proceeding if has a break (or return or continue)

File: build/lex.go [hunk 2]
```diff
@@ -571,13 +632,19 @@ func (in *input) skipPython(p []byte) []byte {
 				rest = p[i:]
 			}
 
-			if !isBlankOrComment(p[i:]) {
-				if !hasPythonContinuation(p[i:]) && c != ' ' && c != '\t' {
-					// Yes, stop here.
-					break
+			if tables.FormatBzlFiles {
+				// In the bzl files mode we only care about the end of the statement, we've found it.
+				break
```
Comment (pmbethe09) [inline build/lex.go]: by my read, this breaks the for loop just to 'return rest'  -- (and also check for quote).
Instead just 'return rest' here, and after the declaration of quote, you can add a 
defer func() {
  if quote != 0 {
 		in.Error("EOF scanning Python quoted string")
 	}
}()

File: build/lex.go [hunk 3]
```diff
@@ -571,13 +632,19 @@ func (in *input) skipPython(p []byte) []byte {
...
+			} else {
+				// In the legacy mode we need to find where the current block ends
+				if !isBlankOrComment(p[i:]) {
+					if !hasPythonContinuation(p[i:]) && c != ' ' && c != '\t' {
+						// Yes, stop here.
+						break
```
Comment (pmbethe09) [inline build/lex.go]: ditto, 'return rest'

**Comment by vladmos** [comment]
Addressed the comments.

This `skipStmt` function existed before (although for different reasons: to skip entire Python blocks, I need 90% of its former functionality to skip statements within which I don't need to track indentation). Once I remove the old behavior it'll be easier to refactor and simplify the code, also in other files.

File: build/lex.go [hunk 4]
```diff
@@ -171,52 +180,66 @@ func (in *input) Lex(val *yySymType) int {
...
-				in.startToken(val)
-				in.readRune()
-				in.endToken(val)
-				return '\n'
-			}
 			if c == '\n' {
+				in.indent = 0
+				in.cleanLine = true
+				if in.endStmt == 0 {
+					// Not in a rule. Tell parser about top-level blank line.
```
Comment (laurentlb) [inline build/lex.go]: Replace 'rule' with 'statement'?

File: build/parse.y [hunk 5]
```diff
@@ -88,6 +89,9 @@ package build
 %token	<pos>	_OR      // keyword or
 %token	<pos>	_PYTHON  // uninterpreted Python block
 %token	<pos>	_STRING  // quoted string
+%token	<pos>	_DEF     // keyword def
+%token  <pos> _INDENT  // indentation
```
Comment (laurentlb) [inline build/parse.y]: fix alignment

File: build/parse.y [hunk 6]
```diff
@@ -108,6 +112,7 @@ package build
 %type	<exprs>		keyvalues_no_comma
 %type	<string>	string
 %type	<strings>	strings
+%type <block>   block
```
Comment (laurentlb) [inline build/parse.y]: alignment

File: build/parse.y [hunk 7]
```diff
@@ -442,6 +457,19 @@ expr:
                         Else: $5,
                 }
 	}
+| _DEF _IDENT '(' exprs_opt ')' ':' block
```
Comment (laurentlb) [inline build/parse.y]: does it support single-line function definition?
def f(): return 1

if not, add a TODO

Comment (vladmos) [reply build/parse.y]: Done.

**Comment by laurentlb** [comment]
LGTM

**Pull request merged**